EDU-709 - splitting export page (#3042)
temporalio/documentation: docs/production-deployment/cloud/gcp-export-gcs.mdx

---
id: gcp-export-gcs
title: Exporting Workflow Event History to GCS
sidebar_label: GCP Export to GCS
description: Export Workflow History to GCS
slug: /cloud/export/gcp-export-gcs
toc_max_heading_level: 4
keywords:
  - explanation
  - how-to
  - operations
  - temporal cloud
  - term
  - verify
  - workflow history export
  - gcp
tags:
  - explanation
  - how-to
  - operations
  - temporal-cloud
  - term
  - verify
  - workflow-history-export
  - gcp
---

> ## **Prerequisites**

Reviewer: I would unbold these? Unless it's for a particular reason but it doesn't match the rest of the docs.

Author: Fixed

Before configuring the Export sink, please complete the following steps in Google Cloud.

1. Create a GCS bucket and take note of its bucket name, for example, "test-export"

>     1. If you wish to enable customer-managed encryption keys (CMEK), do so.

Reviewer: ```suggestion
Enable customer-managed encryption keys (CMEK) if you need additional security for your GCS bucket.
```
something like that

2. Record the GCP Project ID that owns the bucket.

> 3. Set up a service account in the same project that trusts the Temporal internal service account to let Temporal write information to your account.

Reviewer: ```suggestion
Create a service account in the same project that grants Temporal permission to write to your GCS bucket.
```
or something similar for clarity

> 4. Follow the instructions in the Temporal Cloud UI, there are two ways to set up this service account:

Reviewer: This heirachy is fine but probably could be made a little more clearer. Would this be accurate? 
"Follow the instructions in the Temporal Cloud UI to set up the service account. You have two options:

Manual Setup:
1. Input the service account ID, GCP project ID, and GCS bucket name as prompted.
2. Follow the on-screen instructions to manually create a new service account.

Automated Setup:
1. Use the provided [Terraform template](https://github.com/temporalio/terraform-modules/tree/main/modules/gcp-sink-sa) to automatically create the service account.""

Author: Sure thing.

    1. Input the service account ID, GCP project ID and GCS bucket name.
        1. Follow the instructions, manually set up a new service account.
        2. Use the [Terraform template](https://github.com/temporalio/terraform-modules/tree/main/modules/gcp-sink-sa) to create the service account.

## Configure Workflow History Export 

> **How to configure Workflow History Export**

Reviewer: suggestion```
How to configure Workflow History Export using the Temporal Cloud UI or tcld
```

  The exported files will be located in your GCS bucket, following the specified format: 
  ```bash
  //[test-bucket]/temporal-workflow-history/export/[Namespace]/[Year]/[Month]/[Date]/[Hour]/[Minute]/
  ```
3. **UI Updates**: Within the UI, you will find two important pieces of information
    1. **Last Successful Export**: This indicates the date and time when the most recent successful export occurred. 
      It provides an overview of when the export process was last executed successfully.
    2. **Last Status Check**: This refers to an internal Workflow that performs periodic status checks on the Export process. 
      It displays the date and time of the most recent status check, ensuring that the export functionality is continuously monitored.

By periodically reviewing the GCS bucket for exported files and checking the relevant information in the UI, you can stay updated on the progress and status of exports.

## **Working with exported files**

The proto schema is defined in https://github.com/temporalio/api/blob/master/temporal/api/export/v1/message.proto. You could build your own custom deserializer for it. 

> You can find sample go code to reference when you want to deserialize export file in the Appendix section of this document.

Reviewer: Looks great! Main thing is the headers.

Author: Made some changes. Ready for re-review.

---
id: gcp-export-gcs
title: Exporting Workflow Event History to GCS
sidebar_label: GCP Export to GCS
description: Export Workflow History to GCS
slug: /cloud/export/gcp-export-gcs
toc_max_heading_level: 4
keywords:
  - explanation
  - how-to
  - operations
  - temporal cloud
  - term
  - verify
  - workflow history export
  - gcp
tags:
  - explanation
  - how-to
  - operations
  - temporal-cloud
  - term
  - verify
  - workflow-history-export
  - gcp
---

## Prerequisites {#prerequisites}

Before configuring the Export sink, please complete the following steps in Google Cloud.

> 1. Create a GCS bucket and take note of its bucket name, for example, "test-export"
> Enable customer-managed encryption keys (CMEK) if you need additional security for your GCS bucket.
> 2. Record the GCP Project ID that owns the bucket.
> Create a service account in the same project that grants Temporal permission to write to your GCS bucket.

Reviewer 2: I don't think this shows up as you were expecting:
<img width="969" alt="image" src="https://github.com/user-attachments/assets/c6da6304-d5cc-4211-bfb3-b3dba989c8e9">

I think if you want to align things on a separate line, you need to keep the indentation? And add a blank line to make sure it does not get added at the end of the previous line. Alternatively, we could use a sub-dot ?

```suggestion
1. Create a GCS bucket and take note of its bucket name, for example, "test-export"

   Enable customer-managed encryption keys (CMEK) if you need additional security for your GCS bucket.
   
2. Record the GCP Project ID that owns the bucket.

   Create a service account in the same project that grants Temporal permission to write to your GCS bucket.
```

Or something like that...

4. Follow the instructions in the Temporal Cloud UI, there are two ways to set up this service account:
    - Manual Setup:
      - Input the service account ID, GCP project ID and GCS bucket name.
      - Follow the instructions, manually set up a new service account.
    - Automated Setup:

>       - Use the [Terraform template](https://github.com/temporalio/terraform-modules/tree/main/modules/gcp-sink-sa) to create the service account.

Author: We don't typically do numbered lists in sub-bullets cc: @jsundai

## Configure Workflow History Export 

How to configure Workflow History Export using the Temporal Cloud UI or tcld

You can use either the [Temporal Cloud UI](#using-temporal-cloud-ui) or [tcld](#using-tcld) to configure the Workflow History Export.

### Using Temporal Cloud UI

The following steps guide you through setting up Workflow History Export using the Temporal Cloud UI.

![](/img/export-sink-ui-gcp.png)

1. In the Cloud UI, navigate to the Namespaces section. Confirm that the Export feature is visible and properly displayed.
2. Configure the Export sink for a Namespace:
    1. Choose GCS as the Sink type.
    2. Provide the following information:
        1. Name: [User-defined]
        2. Service account ID: [from Prerequisite 3]
        3. GCP Project ID: [from Prerequisite 2]

>         4. GCS bucket name: [from Prerequisite 1]

Reviewer 2: We should probably make that links to anchors for easier navigation ?

Author: I just removed them. They are right above, and the anchors would make it a weird experience. This was just a leftover from a copy paste.

3. After inputting the necessary values, click on **Verify**. 

>   You should be able to write to sink successfully. 

Reviewer 2: ```suggestion
  You should be able to write to the sink successfully. 
```

  If not, please fix any errors or reach out to support for help.
4. Clicking **Create** will complete the Export sink set up
5. The page will auto-refresh and you should see the status “Enabled” on the Export screen. 
  You are now ready to export workflow histories.
6. You can toggle the enable button if you want to stop export and resume in the future. 

>   Note: when you re-enable the feature, it will start from the current point in time, and not from the time when you disable export.

Reviewer 2: ```suggestion
  Note: when you re-enable the feature, it will start from the current point in time, and not from the time when you disabled export.
```

> 7. You can also delete export by clicking Delete.

Reviewer 2: Compared to disabling, does it mean it deletes all export information? If yes, we should probably mention it? (seems logical but...)

Author: I don't know the answer to this one, not as familiar with GCP. Who can we ask to confirm this?

Reviewer 2: I think it's more an export question than GCP?
@alice-yin or @jumbogo might be able to help

    
    OPTIONS:
       --help, -h  show help
    
    ```
    
4. Run the `tcld n export gcs create` command and provide the following information:
    - —namespace
    - —sink-name
    - —service-account-email
    - —gcs-bucket
    - For Example:
    
    ```bash
    tcld n export gcs create -n test.ns --sink-name test-sink --service-account-email test-sink@test-export-sink.iam.gserviceaccount.com --gcs-bucket test-export-validation`
    ```
5. Check the status of this command by either viewing the Namespace Export status in the Temporal Cloud UI or using the following command and looking for the state of “Active”.

```bash
tcld n export gcs g -n test.ns --sink-name test-sink
{
	"name": "test.ns",
	"resourceVersion": "b954de0c-c6ae-4dcc-90bd-3918b52c3f28",
	"state": "Active",
	"spec": {
		"name": "test-sink",
		"enabled": true,
		"destinationType": "Gcs",
		"s3Sink": null,
		"gcsSink": {
			"saId": "test-sink",
			"bucketName": "test-export-validationn",
			"gcpProjectId": "test-export-sink",
		}
	},
	"health": "Ok",
	"errorMessage": "",
	"latestDataExportTime": "0001-01-01T00:00:00Z",
	"lastHealthCheckTime": "2024-01-23T06:40:02Z"
}
```

> ## **Monitoring after Export is enabled**

Reviewer 2: Why is that header setup in bold ? IIRC it does not change anything in docusaurus ?

```suggestion
## Monitoring after Export is enabled
```

1. **Export Job Execution**: The Export job will run as an hourly process, starting at 10 minutes past each hour. 
  This means that closed Workflow history data will be exported to your designated bucket approximately 1 hour after the closure of each Workflow. 
  Please note that the export process may take some time to complete.
2. **Files in GCS Bucket**: After the first hour, you should check your GCS bucket to see if the exported Workflow history files have landed. 
  These files will contain relevant data from closed Workflows. 
  The exported files will be located in your GCS bucket, following the specified format: 

>   ```bash

Reviewer 2: Same as above comment about language

Author: Fixed.

  //[test-bucket]/temporal-workflow-history/export/[Namespace]/[Year]/[Month]/[Date]/[Hour]/[Minute]/
  ```
3. **UI Updates**: Within the UI, you will find two important pieces of information
    1. **Last Successful Export**: This indicates the date and time when the most recent successful export occurred. 
      It provides an overview of when the export process was last executed successfully.
    2. **Last Status Check**: This refers to an internal Workflow that performs periodic status checks on the Export process. 
      It displays the date and time of the most recent status check, ensuring that the export functionality is continuously monitored.

By periodically reviewing the GCS bucket for exported files and checking the relevant information in the UI, you can stay updated on the progress and status of exports.

> ## **Working with exported files**

Reviewer 2: Same, header in bold
```suggestion
## Working with exported files
```
